### PR TITLE
gnash: fix compilation with libgif 5.2

### PR DESCRIPTION
**Pull request by marius851000**
###### Motivation for this change
gnash doesn't compile

###### Things done

Added a patch to gnash that correct the verification for libgif, after it's 5.2 update.

I will try to send the patch upstream.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @rnhmjoj

### CODE REVIEW

**Review by rnhmjoj — APPROVED** [review]
Without the hydra notifications I don't have an easy way to track builds failures.
Anyway, it builds and looks all right. Thank you!

**Review by timokau — CHANGES_REQUESTED** [review]
Please add a short comment that explains why the patch is necessary and ideally also links to an upstream issue.

**Comment by marius851000** [comment]
@timokau There isn't an upstream issue (I will create one), and the patch is mandatory, because the version detection does not work fine.

**Comment by timokau** [comment]
Why doesn't it work fine? Are you sure it wasn't intended to check for 5.1?

In any case, please add a comment explaining this (and a link to the upstream issue after creating it) to the file.

**Comment by marius851000** [comment]
@timokau I did that

**Review by timokau — APPROVED** [review]

**Review by timokau — CHANGES_REQUESTED** [review]
The build fails for me because of a hash mismatch when trying to fetch the patch.

**Comment by marius851000** [comment]
@timokau that work for me:
nix-prefetch-url "https://savannah.gnu.org/patch/download.php?file_id=47859"

[0.0 MiB DL]

path is '/nix/store/zib4ckh512pm5366ab0zg0q2ggp0wc7q-download.php?file_id=47859'

1bwi50qr7wlf8dm3bf7clnly0l69v3nk8gh9h8pdlnhsb52d311j

**Comment by d-goldin** [comment]
Can confirm hash mismatch:

```
100   426  100   426    0     0    432      0 --:--:-- --:--:-- --:--:--   432
hash mismatch in fixed-output derivation '/nix/store/0av5md9blfybkbb6czsv84kd4rmfhn24-download.php?file_id=47859':
  wanted: sha256:1bwi50qr7wlf8dm3bf7clnly0l69v3nk8gh9h8pdlnhsb52d311j
  got:    sha256:0aimayzgi5065gkcfcr8d5lkd9c0471q7dqmln42hjzq847n6d5y
```

@marius851000: Keep in mind that if you used `nix-prefetch-url` or something like that, you'd have a cached version in your store, but fetchpatch normalized the patch and so the checksum actually turns out to be different.

**Comment by rnhmjoj** [comment]
I get the same hash as @marius851000. This is pretty strange.

**Comment by d-goldin** [comment]
> I get the same hash as @marius851000. This is pretty strange.

With `nix-prefetch-url` or actually running it through fetchpatch? See my remark above regarding `nix-prefetch-url` - its a common pitfall with `fetchpatch`;

```
> nix repl '<nixpkgs>'

nix-repl> :b fetchpatch { url = "https://savannah.gnu.org/patch/download.php?file_id=47859"; sha256 = lib.fakeSha256; }
hash mismatch in fixed-output derivation '/nix/store/0av5md9blfybkbb6czsv84kd4rmfhn24-download.php?file_id=47859':
  wanted: sha256:0000000000000000000000000000000000000000000000000000
  got:    sha256:0aimayzgi5065gkcfcr8d5lkd9c0471q7dqmln42hjzq847n6d5y
[0 built (1 failed), 1 copied (0.2 MiB), 0.1 MiB DL]
error: build of '/nix/store/p1nl81nks7943yyr2m61q4kkkcw6g090-download.php?file_id=47859.drv' failed
```

File: pkgs/misc/gnash/default.nix [hunk 1]
```diff
@@ -93,6 +93,13 @@ stdenv.mkDerivation {
     ++ optionals enablePlugins   [ xulrunner npapi_sdk ]
     ++ optionals enableGTK       [ gtk2 gnome2.gtkglext gnome2.GConf ];
 
+  patches = [
+    (fetchpatch { # fix compilation due to bad detection of libgif version: https://savannah.gnu.org/patch/index.php?9873
+      url = "https://savannah.gnu.org/patch/download.php?file_id=47859";
```
Comment (jonringer) [inline pkgs/misc/gnash/default.nix]: I got a redirect to this URL, may help with the hash mismatch
```suggestion
      url = "https://savannah.gnu.org/file/DGifCloseFile-update.patch?file_id=47859";
```

Comment (jtojnar) [reply pkgs/misc/gnash/default.nix]: Redirects do not affect hash mismatches (unless they change to point to a different file). The only thing that matters for fixed output derivations is the final content. Here, the output after passing the patch through `filterdiff`, so the hash will not match the one of unprocessed patch as described above.

**Comment by marius851000** [comment]
@timokau I fixed that. At least, I now know what fetchpath does.

**Comment by rnhmjoj** [comment]
@d-goldin You were right, I was using prefetch-url.

**Comment by d-goldin** [comment]
@GrahamcOfBorg build gnash

**Comment by d-goldin** [comment]
> @timokau I fixed that. At least, I now know what fetchpath does.

Seems to still be the old/wrong hash and fails the checks: https://github.com/NixOS/nixpkgs/pull/73040/checks?check_run_id=311777285

**Comment by marius851000** [comment]
@timokau You're right, sorry. I just haven't added my change to git.

**Review by timokau — APPROVED** [review]

**Comment by timokau** [comment]
Thanks!

**Pull request merged**